src/Sparrow.Server/Collections/SmallSet.cs

```diff
@@ -0,0 +1,281 @@
+                    chunk = new Vector<TKey>(keys, elementIdx + 1);
+                    chunk = Vector.Equals(keyVector, chunk);
+                    if (chunk == Vector<TKey>.Zero)
+                        continue;
+
+                    elementIdx = Math.Min(elementIdx + Vector256<TKey>.Count, _length - 1);
+                    goto Found; 
+                }
+
+                chunk = new Vector<TKey>(keys);
+                chunk = Vector.Equals(keyVector, chunk);
+                if (chunk == Vector<TKey>.Zero)
+                    return Invalid;
+
+                elementIdx = Vector256<TKey>.Count - 1;
+
+                Found:
+                elementIdx = Math.Min(elementIdx, _currentIdx);
+            }
+
+            while (elementIdx >= 0)
+            {
+                ref TKey current = ref keys[elementIdx];
+                if (current.Equals(key))
+                    return elementIdx;
+
+                elementIdx--;
+            }
+
+            return Invalid;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int RequestWritableBucket()
+        {
+            _currentIdx++;
+            return _currentIdx % _length;
+        }
+
+        public bool TryGetValue(TKey key, out TValue value)
+        {
+            int idx = FindKey(key);
+            if (idx == Invalid)
+            {
+                Unsafe.SkipInit(out value);
+                return false;
+            }
+                
+            value = _values[idx];
+            return true;
+        }
+
+        public void Clear()
+        {
+            Array.Fill(_keys, (TKey)(object)-1);
+            Array.Fill(_values, default);
+            _currentIdx = -1;
+        }
+
+        public void Dispose()
+        {
+            ArrayPool<TKey>.Shared.Return(_keys);
+            ArrayPool<TValue>.Shared.Return(_values);
+        }
+    }
+
+    // The concept of the small set is about a normal dictionary optimized for accessing recently accessed items. 
+    // In a sense it behaves like a LRU cache over a dictionary, however, it will not use the backing dictionary unless it has to.
+    // This version of the Small Set is specially designed to deal with blittable keys in a very efficient manner.
+    // Since the scanning of the set is based on SIMD instructions (unless they are not available on the platform)
+    // the cost is linear on the size and may become too high if there are too many novel accesses.
+    public sealed class SmallSet<TKey, TValue> : IDisposable
+        where TKey : unmanaged
+    {
+        private const int Invalid = -1;
+
+        private readonly int _length;
+        private readonly TKey[] _keys;
+        private readonly TValue[] _values;
+        private Dictionary<TKey, TValue> _overflowStorage;
+        private int _currentIdx;
+
+        public SmallSet(int size = 0)
+        {
+            _length = size > Vector<TKey>.Count ? size - size % Vector<TKey>.Count : Vector<TKey>.Count;
```

Comment:
`size - size % Vector<TKey>.Count `

Add `()` here, off the top of my head, not sure what is the priority order, and we had similar bugs